### PR TITLE
add inference_mode context manager with decorator support

### PR DESCRIPTION
**Pull request by MaxWolf-01**
This is https://github.com/tinygrad/tinygrad/pull/3281 but with the right API.

Introduces `inference_mode` context manager akin to [torch.inference_mode](https://pytorch.org/docs/stable/generated/torch.inference_mode.html):
```python
with Tensor.inference_mode(): ...
# OR
@Tensor.inference_mode()
def f(): ...
```
Also adds ability to use `Tensor.train()` as a decorator:
```python
@Tensor.train():
def f(): ...
```

### CODE REVIEW

**Comment by geohot** [comment]
Why is it `val` and not `mode` to match torch?

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -67,12 +68,16 @@ class Tensor:
   __slots__ = "lazydata", "requires_grad", "grad", "_ctx"
   __deletable__ = ('_ctx',)
   training: ClassVar[bool] = False
-  class train:
-    def __init__(self, val=True): self.val = val
-    def __enter__(self): self.prev, Tensor.training = Tensor.training, self.val
+  class train(ContextDecorator):
+    def __init__(self, mode:bool = True): self.mode = mode
```
Comment (geohot) [inline tinygrad/tensor.py:72]: Err, why did this one change? Does this match torch?

Comment (MaxWolf-01) [reply tinygrad/tensor.py:72]: Yes, torch also uses `mode` for `train`. See [source code](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module)

**Comment by geohot** [comment]
Cool, merged

**Pull request merged**